src/crush/CrushWrapper.cc

```diff
@@ -1647,12 +1647,22 @@ int CrushWrapper::update_device_class(int id,
                                       const string& name,
                                       ostream *ss)
 {
+  assert(item_exists(id));
+  auto old_class_name = get_item_class(id);
+  stringstream ts;
+  if (old_class_name &&
+      old_class_name != class_name && // so command is idempotent
+      class_is_in_use(get_class_id(old_class_name), &ts)) {
+    *ss << "osd." << id << " has already bound to class '" << old_class_name
+        << "', which is currently " << ts.str();
+    return -EBUSY;
```

Comment:
It's not if a CRUSH rule points to it.  I see that you're trying to prevent the situation where that happens (a crush rule references a class with no devices), but it might be simpler to just allow it.  e.g., generate shadow hierarchies for any (class,root) combination referenced by a crush rule.  If the users removes the last device in the class the rule will point to an empty shadow root and return no device, and those PGs will become remapped (probably) or some other semi-broken PG state until they fix their crush rule.  That seems more friendly and intuitive than being prevented from changing a device's class until you delete the crush rule.